### PR TITLE
[new-model] Add native Helios-Distilled T2V pipeline

### PR DESCRIPTION
**Pull request by FlappyBob**
## Purpose

Add native, runnable FastVideo support for `BestWishYsh/Helios-Distilled`
text-to-video generation. This expands the original transformer-only PR with
the scheduler, stage-composed pipeline, registry/preset wiring, typed inference
example, component parity, pipeline parity, and real-video evidence requested
during review.

## Supported Scope

- Helios-Distilled T2V;
- official Diffusers-layout checkpoint at revision
  `1999182614cb08d3bdcc46b9827504af2914b87b`;
- native Helios transformer and DMD scheduler;
- native FastVideo pipeline using the existing Wan VAE, UMT5 encoder, and
  tokenizer loader after exact-asset parity.

Base/Mid checkpoints, `transformer_ode`, training, ODE workflows, I2V, and
conditioned generation are outside this PR.

## Changes

- add `HeliosConfig` and native `HeliosTransformer3DModel` with direct
  1101/1101 checkpoint-key compatibility;
- add native `HeliosDMDScheduler` and scheduler registry resolution;
- add `HeliosPipelineConfig`, `helios_distilled_t2v` preset, exact model
  detection, and `HeliosPyramidPipeline` entry-class resolution;
- compose validation, shared text encoding/conditioning, Helios-specific
  autoregressive pyramid denoising, and chunk decoding stages;
- expose Helios sampling controls through `SamplingParam`, typed
  `SamplingConfig`, CLI parsing, schema inventory, and `ForwardBatch`;
- add a typed public inference example at
  `examples/inference/basic/basic_helios_distilled_t2v.py`;
- add component, stage, smoke, end-to-end latent, and real-video integrity
  tests plus reviewer setup/state records.

The worker moves final decoded/latent outputs to CPU before multiprocessing
return. This prevents CUDA-IPC allocation OOM after high-memory inference and
is protected by a regression test.

## Pinned References

- architecture: `PKU-YuanGroup/Helios@8f2a2faab3298c8a7630a2c73aea37c01b5bab01`;
- executable parity reference: Diffusers `0.39.0`;
- checkpoint: `BestWishYsh/Helios-Distilled@1999182614cb08d3bdcc46b9827504af2914b87b`.

### Components

```text
Transformer: 24 passed, 0 skipped
  strict load: 1101/1101
  full BF16 normal abs-mean drift: 0.0136%
  full BF16 pyramid abs-mean drift: 0.0305%
  tiny normal/pyramid: exact
  FlashAttention-vs-SDPA: exact
  SP=2 max/mean diff: 0.00000036 / 0.00000005

VAE + UMT5/tokenizer + scheduler: 17 passed, 0 skipped
  VAE decode max/mean diff: 0 / 0
  UMT5 FP32 max/mean diff: 0.00000125 / 0.00000010
  UMT5 BF16 max/mean diff: 0.02343750 / 0.00160135
  scheduler stages/steps: bit-exact
```

### Pipeline

```text
math/stage/smoke: 29 passed
end-to-end latent parity: 1 passed in 82.85 s
shape: [1, 16, 9, 16, 24]
cosine: 0.97970295
abs-mean drift: 0.3763%
mean absolute diff: 0.17001711
RMSE: 0.24650776
```

### Public Example

```text
H.264, 640x384, 24 FPS, 33 frames, 1.375 s
generation: 29.43 s
full ffmpeg decode: passed
container/non-black-frame gate: passed
SHA-256: 2e6d5099715256b16b865cb300479b328b2835f9f8792cb5427a98c17eb0b038
```

### Static Checks

```text
pre-commit run --all-files: all hooks passed
```

## CI Base Alignment

The branch is rebased onto `upstream/main@a159b63c`. The current shared
`.buildkite/scripts/unit_test.sh` passes all 1047 collected tests locally in
28.70 seconds. `pre-commit run --all-files` also passes.

## Quality Regression

Local real-video integrity and non-black-frame checks pass. Publishing a CI
SSIM reference is deferred because reference upload requires separate approval;
this PR does not claim CI SSIM passed.

## Checklist

- [x] Native transformer and scheduler
- [x] Reused-component non-skip parity
- [x] Runnable stage-composed pipeline and preset
- [x] Typed API and inference example
- [x] Pipeline smoke and non-skip latent parity
- [x] Valid decoded video output
- [x] `pre-commit run --all-files`
- [x] Rebased onto latest main; shared unit script passes 1047 tests
- [ ] Maintainer review and approval
- [ ] CI SSIM reference publication, if requested and separately approved

### CODE REVIEW

File: fastvideo/pipelines/basic/helios/stages.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+
+        latent_chunks = batch.helios_latent_chunks
+        if not isinstance(latent_chunks, list) or not latent_chunks:
+            raise ValueError("Helios decoding requires non-empty helios_latent_chunks")
+        decoded_chunks = [self.decode(chunk, fastvideo_args) for chunk in latent_chunks]
+        frames = torch.cat(decoded_chunks, dim=2)
+
+        temporal_scale = fastvideo_args.pipeline_config.vae_config.arch_config.scale_factor_temporal
+        generated_frames = get_generated_pixel_frames(frames.shape[2], temporal_scale)
+        batch.output = frames[:, :, :generated_frames].detach().to(dtype=torch.float32, device="cpu")
```
[the remark this answers is not in the document]
Comment (FlappyBob) [reply fastvideo/pipelines/basic/helios/stages.py:390]: Fixed in 3238ae70. The chunk decoder now caps decoded output at `batch.num_frames`, and the regression test requests 34 frames from two chunks and asserts an exact 34-frame result.

File: fastvideo/configs/pipelines/helios.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+def helios_postprocess_text(output: BaseEncoderOutput) -> torch.Tensor:
+    if output.last_hidden_state is None or output.attention_mask is None:
+        raise ValueError("Helios UMT5 output requires hidden states and attention mask")
+    hidden_states = output.last_hidden_state
+    sequence_lengths = output.attention_mask.gt(0).sum(dim=1).long()
+    if torch.isnan(hidden_states).any():
+        raise ValueError("Helios UMT5 produced NaN hidden states")
+    trimmed = [hidden[:length] for hidden, length in zip(hidden_states, sequence_lengths, strict=True)]
+    return torch.stack(
+        [torch.cat([hidden, hidden.new_zeros(512 - hidden.shape[0], hidden.shape[1])]) for hidden in trimmed],
```
[the remark this answers is not in the document]
Comment (FlappyBob) [reply fastvideo/configs/pipelines/helios.py:51]: Fixed in 3238ae70. Helios text postprocessing now truncates valid tokens to the fixed 512-token model context before padding, with a 520-token regression test.

File: fastvideo/pipelines/basic/helios/stages.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        if not isinstance(latent_chunks, list) or not latent_chunks:
+            raise ValueError("Helios decoding requires non-empty helios_latent_chunks")
+        decoded_chunks = [self.decode(chunk, fastvideo_args) for chunk in latent_chunks]
+        frames = torch.cat(decoded_chunks, dim=2)
+
+        temporal_scale = fastvideo_args.pipeline_config.vae_config.arch_config.scale_factor_temporal
+        assert isinstance(batch.num_frames, int)
+        generated_frames = min(
+            batch.num_frames,
+            get_generated_pixel_frames(frames.shape[2], temporal_scale),
```
Comment (H1yori233) [inline fastvideo/pipelines/basic/helios/stages.py]: [P2] Preserve requested frames at complete-chunk boundaries

The new min() fixes the 34-frame overrun but still drops frames that have already been decoded: requesting 66 frames creates two 33-frame chunks, then get_generated_pixel_frames(66, 4) rounds the available output down to 65. Likewise 99 returns 97, 132 returns 129, and 264 returns 261. Input validation accepts these positive frame counts, so the returned video is shorter than the request even though enough frames exist. Crop directly to batch.num_frames (bounded by frames.shape[2] if needed), without applying the latent temporal-alignment formula again to concatenated decoded frames, and add regression cases for these boundaries. Confirmed by executing this decoder method on CPU with stub 33-frame chunk outputs.

**Review by H1yori233 — DISMISSED** [review]

File: fastvideo/configs/models/dits/helios.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: Apache-2.0
+from dataclasses import dataclass, field
+
+from fastvideo.configs.models.dits.base import DiTArchConfig, DiTConfig
+from fastvideo.platforms import AttentionBackendEnum
+
+
+def _is_transformer_block(name: str, module) -> bool:
+    del module
+    return name.startswith("blocks.") and name.split(".")[-1].isdigit()
```
Comment (H1yori233) [inline fastvideo/configs/models/dits/helios.py]: [P1] Restrict FSDP matching to complete transformer blocks

This predicate also matches `blocks.0.attn2.to_out.0` (plus the other numbered descendants), so `shard_model()` independently wraps the cross-attention output projection in FSDP. With `sp_size=2`, the history-only shard has `current_count == 0` and skips `attn2`, while the other rank enters that sharded projection. The ranks consequently execute different FSDP collectives and generation hangs. On two A100s, the actual 14B checkpoint stalls in its first 128×192×33 chunk with `use_fsdp_inference=True`; a two-layer tiny reproduction also times out. Keeping all model computation unchanged and restricting this predicate to exact `blocks.<integer>` names makes both the tiny test and real 14B two-GPU inference complete normally (real generation: 11.05 seconds). Please use an exact block-name/type predicate and cover FSDP+SP with a history-only rank in a regression test. SP-only parity does not exercise this failure.

**Review by H1yori233 — DISMISSED** [review]

**Comment by H1yori233** [comment]
Sorry, my agent kind of went rogue 😅...

**Comment by H1yori233** [comment]
I found that inference hangs with FSDP + SP on 2 GPUs. Could you fix this by matching only top-level transformer blocks, following [LingBot-Video](https://github.com/hao-ai-lab/FastVideo/blob/3238ae70ffc8caf628326eec37590120dd31630f/fastvideo/configs/models/dits/lingbot_video.py#L12-L16)?

**Comment by FlappyBob** [comment]
> I found that inference hangs with FSDP + SP on 2 GPUs. Could you fix this by matching only top-level transformer blocks, following [LingBot-Video](https://github.com/hao-ai-lab/FastVideo/blob/3238ae70ffc8caf628326eec37590120dd31630f/fastvideo/configs/models/dits/lingbot_video.py#L12-L16)?

Fixed in c22f6897.
